src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/ThrowsTaglet.java

```diff
@@ -47,252 +52,733 @@
-        // TODO: assert declaredExceptionTypes are instantiated
-        var utils = writer.configuration().utils;
-        Content result = writer.getOutputInstance();
-        for (TypeMirror declaredExceptionType : declaredExceptionTypes) {
-            TypeElement te = utils.asTypeElement(declaredExceptionType);
-            if (te != null &&
-                    !alreadyDocumented.contains(declaredExceptionType.toString()) &&
-                    !alreadyDocumented.contains(utils.getFullyQualifiedName(te, false))) {
-                if (alreadyDocumented.isEmpty()) {
-                    result.add(writer.getThrowsHeader());
-                }
-                result.add(writer.throwsTagOutput(declaredExceptionType));
-                alreadyDocumented.add(utils.getSimpleName(te));
+    private static AssertionError newAssertionError(Object... objects) {
+        return new AssertionError(Arrays.toString(objects));
+    }
+
+    private static String diagnosticDescriptionOf(Element e) {
+        var name = e instanceof QualifiedNameable q ? q.getQualifiedName() : e.getSimpleName();
+        return name + " (" + detailedDescriptionOf(e) + ")";
+    }
+
+    private static String detailedDescriptionOf(Element e) {
+        // It might be important to describe the element in detail. Sometimes
+        // elements share the same simple and/or qualified name. Outputting
+        // individual components of that name as well as their kinds helps
+        // the user disambiguate such elements.
+        // TODO: is it feasible to use javax.lang.model.util.Elements.getBinaryName(TypeElement)
+        //  to augment/replace detailed information for _type elements_?
+        //  (binary names don't work for _type variables_)
+        var thisElementDescription = e.getKind() + " " + switch (e.getKind()) {
+            // A package is never enclosed in a package and a module is
+            // never never enclosed in a module, no matter what their
+            // qualified name might suggest. Get their qualified
+            // name directly. Also, unnamed packages and
+            // modules require special treatment.
+            case PACKAGE -> {
+                var p = (PackageElement) e;
+                // TODO: i18n
+                yield p.isUnnamed() ? "<unnamed package>" : p.getQualifiedName();
+            }
+            case MODULE -> {
+                var m = (ModuleElement) e;
+                // TODO: i18n + is there any value is displaying unnamed module?
+                yield m.isUnnamed() ? "<unnamed module>" : m.getQualifiedName();
             }
+            default -> e.getSimpleName();
+        };
+        if (e.getEnclosingElement() == null) {
+            return thisElementDescription;
         }
-        return result;
+        var enclosingElementDescription = detailedDescriptionOf(e.getEnclosingElement());
+        return enclosingElementDescription + " " + thisElementDescription;
```

Comment:
A named package is enclosed by a module, although it may be the unnamed module for >= JDK 9.
A well-formed named module never has an unnamed package.

Consider return empty string from the switch and dealing with that in the `return` statement.

That being said, who uses this info where: you're including the raw element kind in the result, so is this mostly for debugging?

Reply:
It will need to be better stringified and i18n-zed. That output is not for debugging, it's for diagnosing.